### ISSUE TITLE
RPP not convergence

### ISSUE THREAD

**Issue by yja1** [YHJRUBY/ReID_baseline#4]
I try to do RPP part. but not converge.

`class PCBRPPModel(nn.Module):
    def __init__(self, num_classes, num_stripes=6, local_conv_out_channels=256, train_stage = "PCB"):

        super(PCBRPPModel, self).__init__()
        self.num_stripes = num_stripes
        self.num_classes = num_classes
        self.features_G = []
        self.features_H = []
        self.train_stage = train_stage #####-----add for change train stage

        resnet = models.resnet50(pretrained=True)
        # Modifiy the stride of last conv layer
        resnet.layer4[0].conv2 = nn.Conv2d(
            512, 512, kernel_size=3, bias=False, stride=1, padding=1)
        resnet.layer4[0].downsample = nn.Sequential(
            nn.Conv2d(1024, 2048, kernel_size=1, stride=1, bias=False),
            nn.BatchNorm2d(2048))

        # Remove avgpool and fc layer of resnet
        modules = list(resnet.children())[:-2]
        self.backbone = nn.Sequential(*modules)

        #####-----add for RPP
        #####-----fc method
        """self.fc_seg6 = nn.Linear(2048, num_stripes, bias=False)
        nn.init.normal(self.fc_seg6.weight, std=0.001)
        self.avgpool_seg6 = nn.AdaptiveAvgPool2d((1,1))
        """
        #####-----
        #####-----1x1conv method
        self.conv1x1_seg6 = nn.Conv2d(2048, num_stripes, 1, bias=False)
        self.sm_seg6 = nn.Softmax2d()
        #####-----add for RPP

        # Add new layers
        self.avgpool = nn.AdaptiveAvgPool2d((self.num_stripes, 1))
        self.local_conv = nn.Sequential(
            nn.Conv2d(2048, local_conv_out_channels, 1),
            nn.BatchNorm2d(local_conv_out_channels),
            nn.ReLU(inplace=True))

        # Classifier for each stripe
        self.fc_list = nn.ModuleList()
                for _ in range(num_stripes):
            fc = nn.Sequential(
                nn.Linear(local_conv_out_channels, num_classes))
            # nn.Softmax())

            # fc initialize
            nn.init.normal(fc[0].weight, std=0.001)
            nn.init.constant(fc[0].bias, 0)

            self.fc_list.append(fc)

    def forward(self, x):
        features = self.backbone(x)

        # [N, C, H, W]
        assert features.size(2) % self.num_stripes == 0, 'Image height cannot be divided by num_strides'

        if self.train_stage == "PCB":
            self.features_G = self.avgpool(features)
        elif self.train_stage == "PCBRPP" or self.train_stage == "RPP":
            #####-----add for RPP

            f_conv1x1_seg6 = self.conv1x1_seg6(features)
            f_sm_seg6 = self.sm_seg6(f_conv1x1_seg6)
            f_chunk_seg6 = torch.chunk(f_sm_seg6, self.num_stripes, 1)
            f_batch = features.size(0)
            features_seg6 = features*f_chunk_seg6[0]+features*f_chunk_seg6[1]+features*f_chunk_seg6[2]+features*f_chunk_seg6[3]+features*f_chunk_seg6[4]+features*f_chunk_seg6[5]

            self.features_G = self.avgpool(features_seg6)#####-----add for RPP
            #####-----add for RPP

        ###########################################
        # [N, C=256, H=S, W=1]
        self.features_H = self.local_conv(self.features_G)

        # H=S * [N, num_classes]
        predicitions_list = []
        for i in range(self.num_stripes):
            local_feature = self.features_H[:, :, i, :].contiguous()
            local_feature = local_feature.view(local_feature.size(0), -1)
            predicitions_list.append(self.fc_list[i](local_feature))

        return predicitions_list

`

I train use lr like this two method:
optimizer = optim.SGD([
    {'params': model.backbone.parameters(), 'lr': 0 },
    {'params': model.local_conv.parameters(), 'lr': 0 },
    {'params': model.fc_list.parameters(), 'lr': 0. },
    {'params' : model.conv1x1_seg6.parameters(), 'lr' : 0.1 }
], momentum=0.9, weight_decay=5e-4, nesterov=True)

or 
optimizer = optim.SGD([
    {'params': model.backbone.parameters(), 'lr': 0.001 },
    {'params': model.local_conv.parameters(), 'lr': 0.001 },
    {'params': model.fc_list.parameters(), 'lr': 0.001 },
    {'params' : model.conv1x1_seg6.parameters(), 'lr' : 0.1 }
], momentum=0.9, weight_decay=5e-4, nesterov=True)

**Comment by biubug6** [comment]
I think your implementation has some problem. Do you have any other idea?

**Comment by ye-hongjiang** [comment]
Have you solved the problem? I don't plan to implement the RPP. Closed.
